engine/table/src/main/java/io/deephaven/engine/table/impl/UpdateBy.java

```diff
@@ -1,96 +1,842 @@
+                    offsetList.add(ii);
+                    countList.add(Math.min(bucketsPerTask, dirtyBuckets.length - ii));
+                }
+
+                jobScheduler.iterateParallel(ExecutionContext.getContextToRecord(), this, 0, offsetList.size(), idx -> {
+                    final int bucketOffset = offsetList.get(idx);
+                    final int bucketCount = countList.get(idx);
+                    for (int bucketIdx = bucketOffset; bucketIdx < bucketOffset + bucketCount; bucketIdx++) {
+                        UpdateByBucketHelper bucket = dirtyBuckets[bucketIdx];
+                        bucket.assignInputSources(winIdx, maybeCachedInputSources);
+                        bucket.processWindow(winIdx, initialStep);
+                    }
+                }, completedAction, this::onError);
+            } else {
+                // minimize overhead when running serially
+                for (UpdateByBucketHelper bucket : dirtyBuckets) {
+                    bucket.assignInputSources(winIdx, maybeCachedInputSources);
+                    bucket.processWindow(winIdx, initialStep);
+                }
+                completedAction.run();
+            }
+        }
+
+        /**
+         * Process all {@code windows} in a serial manner (to minimize cache memory usage). Will create cached input
+         * sources, process the buckets, then release the cached columns before starting the next window. Calls
+         * {@code completedAction} when the work is complete
+         */
+        private void processWindows(final Runnable completeAction) {
+            jobScheduler.iterateSerial(ExecutionContext.getContextToRecord(), this, 0, windows.length,
+                    (winIdx, windowComplete) -> {
+                        UpdateByWindow win = windows[winIdx];
+
+                        // this is a chain of calls: cache, then shift, then process the dirty buckets for this window
+                        cacheInputSources(winIdx, () -> {
+                            // prepare each operator for the parallel updates to come
+                            if (initialStep) {
+                                // prepare the entire set of rows on the initial step
+                                try (final RowSet changedRows = redirHelper.isRedirected()
+                                        ? RowSetFactory.flat(redirHelper.requiredCapacity())
+                                        : source.getRowSet().copy()) {
+                                    for (UpdateByOperator op : win.getOperators()) {
+                                        op.prepareForParallelPopulation(changedRows);
+                                    }
+                                }
+                            } else {
+                                // get the minimal set of rows to be updated for this window
+                                try (final WritableRowSet windowRowSet = shiftedRows.copy()) {
```

Comment:
Let's discuss.

Reply:
Added a clarifying comment for point 1 and discussed point 2-3 in depth .The current implementation includes shifts and modifies in the PPP call, but not removes.  This isn't a problem because we aren't removing values from the output columns so downstream tables will read the correct `prev` values from the current.  When a removed item is replaced by an add or a shift, then it gets included in the next PPP step and its prev value will still be correct.

We should attempt to cleanup our object columns however, so that means we should include removes in PPP and cleanup objects as in `io.deephaven.engine.table.impl.SelectOrUpdateListener#onUpdate` & `io.deephaven.engine.table.impl.select.analyzers.SelectColumnLayer#clearObjectsAtThisLevel`

Reply:
Small scope creep planned.  Split functionality of PPP to ensureBlock / ensurePrevious allowing two non-overlapping RowSets.  This will prevent the unnecessary creation of `prevBlocks` for new items.

Reply:
Created issue #3394